### PR TITLE
#164797147 Ft enable social login with facebook

### PR DESCRIPTION
**Pull request by musonant**
#### What does this PR do?
Enable authentication with facebook

#### Description of Task to be completed?
Create facebook developer application
Implement login with data returned after facebook authentication

#### Any background context you want to provide?
NA

#### What are the relevant pivotal tracker stories?(if applicable)
#164797147

#### Screenshots
Screenshot 2019-04-08 at 10 12 51 AM
Screenshot 2019-04-08 at 10 12 55 AM
Screenshot 2019-04-08 at 10 13 03 AM

### CODE REVIEW

File: src/db/migrations/2019-001-create-users.js [hunk 1]
```diff
@@ -22,12 +22,12 @@ export default {
     },
     email: {
       type: Sequelize.STRING,
-      allowNull: false,
+      allowNull: true,
```
Comment (Japhethca) [inline src/db/migrations/2019-001-create-users.js]: why this change?

Comment (musonant) [reply src/db/migrations/2019-001-create-users.js]: @Japhethca Some facebook users do not have an email attached. For the direct login, validations would prevent signup when email is missing.

File: src/routes/index.js [hunk 2]
```diff
@@ -45,4 +46,7 @@ const router = Router();
 router.post('/users', validateNewUser, createUser);
 router.get('/users/confirmation/:token', confirmUser);
 
+router.get('/auth/facebook', passport.authenticate('facebook', { scope: ['email'] }));
```
Comment (Japhethca) [inline src/routes/index.js:49]: Please add documentation for this routes.

File: src/db/migrations/2019-001-create-users.js (hunk 1 above)
Comment (Japhethca) [reply src/db/migrations/2019-001-create-users.js]: make sense!

File: src/routes/services/passport-strategies.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+dotenv.config();
+
+const { User } = models;
+const FacebookStrategy = require('passport-facebook').Strategy;
+
+const credentials = {
+  facebook: {
+    clientID: process.env.FACEBOOK_APP_CLIENT_ID,
+    clientSecret: process.env.FACEBOOK_APP_SECRET,
+    callbackURL: 'http://localhost:3000/api/v1/auth/facebook/callback',
```
Comment (Japhethca) [inline src/routes/services/passport-strategies.js]: you should also create an environment variable for this, as this will not work when pushed to Heroku.

**Comment by Japhethca** [comment]
Please add a screenshot showing that this is working.

File: src/routes/services/passport-strategies.js (hunk 3 above)
Comment (musonant) [reply src/routes/services/passport-strategies.js]: Yes. I have added the environmental variables to Travis and have asked @NedyUdombat to add it also to the heroku app

File: src/routes/index.js (hunk 2 above)
Comment (musonant) [reply src/routes/index.js:49]: Oh, yeah.

File: src/routes/services/passport-strategies.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+};
+
+const facebookAuth = async (accessToken, refreshToken, profile, done) => {
+  try {
+    const currentUser = await User.findOrCreate({
+      where: { socialId: profile.id },
+      defaults: {
+        firstName: profile.name.givenName,
+        lastName: profile.name.familyName,
+        username: profile.emails[0].value,
```
Comment (Japhethca) [inline src/routes/services/passport-strategies.js:26]: Why are we giving username the value of email?

Comment (musonant) [reply src/routes/services/passport-strategies.js:26]: The username of the facebook user might not be unique on our application. The default username is now set to be the email, since emails are unique.

File: src/routes/controllers/users.js [hunk 5]
```diff
@@ -52,3 +51,27 @@ export const confirmUser = async (req, res) => {
...
+
+/**
+ * Redirect location after social login
+ * @param {object} req
+ * @param {object} res
+ * @returns {object} response object
+ */
+export const socialController = async (req, res) => {
+  const user = req.user[0];
+  const isCreated = req.user[1];
```
Comment (Japhethca) [inline src/routes/controllers/users.js]: we can refactor line 62 and 63 to:
`const [user, isCreated] = req.user;`

File: src/routes/services/passport-strategies.js (hunk 3 above)
Comment (Japhethca) [inline src/routes/services/passport-strategies.js]: we can create an environment variable to hold this url.

**Review by Japhethca — APPROVED** [review]
